orttraining/orttraining/training_api/training_session.cc

```diff
@@ -9,19 +9,21 @@ namespace api {
 
 TrainingSession::TrainingSession(const Environment& session_env,
                                  const SessionOptions& session_options,
+                                 const std::vector<std::shared_ptr<IExecutionProvider>>& providers,
                                  const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters)
     : environment_(session_env),
       session_options_{session_options},
+      providers_{providers},
```

Comment:
why does training session need to maintain a list of providers? All it should do is pass it on so that module and optimizer can use it to register these eps in respective inference sessions.

Reply:
I think this is true for all private members of `TrainingSession` (session_options, session_env, parameters). Really TrainingSession does not need to hold onto any of them.
My intent was to keep the `providers_` and `session_options_` together since they are a common entity in the c_api public `SessionOptions`.
The other reason is that I see `TrainingSession.Initialize` a way to pass initialize the models as opposed to setting the configuration (session_options) on the session. I think either way is fine.

Having said that, if we feel that it does not make sense for training session to hold onto the providers, I can pass it as an argument to `Initialize` and forward it to the `Module` and `Optimizer`.